src/Server/TCPHandler.cpp

```diff
@@ -366,7 +366,10 @@ void TCPHandler::runImpl()
         try
         {
             /// If a user passed query-local timeouts, reset socket to initial state at the end of the query
-            SCOPE_EXIT({state.timeout_setter.reset();});
+            SCOPE_EXIT({
+                std::scoped_lock lock(in_mutex, out_mutex);
+                state.timeout_setter.reset();
```

Comment:
May be better to just move this `state.timeout_setter.reset()` after this try-catch block and after `state.io.onException();` in the catches? When we would do `state.io.onException()` - we reset the pipeline so the race would not longer happen (storageS3cluster reading threads will be finished on pipeline reset). I guess this race happens only in case when exception is thrown in the `try` code? Because otherwise we will not exit the `try` code unless pipeline is finished, e.g. the race would not happen as well.

Reply:
Please correct me if I'm wrong on the next assumptions.
Technically data race condition happens whenever data is written concurrently without either explicit or implicit memory barrier. It doesn't depend on actual events of concurrency - it's enough to have a concurrent code running on different threads. Thread sanitizer detects actual events, not potential condition - because of this ad-hoc solution which can fix triggered thread sanitizer, do not solve the issue of race condition itself - some other piece of code can easily retrigger sanitizer.
So, regarding this particular case, if these assumptions are correct, maybe it worth fixing the general issue and not a special scenario.

Reply:
My proposal also fixes the general case scenario. `in` cannot be used concurrently, but it can now be used by either this TCPHandler thread or by callbacks we set inside this try block. So this is where we have potential data race. But if we finish the pipeline (for example, by calling state.io.onException() in case of exception or by exhausting it by reaching isCompleted) then such race is not possible to happen. There is no other possible data race case on `in` AFAIK.